realsense2_camera/src/tfs.cpp

```diff
@@ -168,12 +168,36 @@ void BaseRealSenseNode::calcAndAppendTransformMsgs(const rs2::stream_profile& pr
 
     // publish static TF
     auto Q = rotationMatrixToQuaternion(tf_ex.rotation);
+
+    // Representing Rotations with Quaternions
+    // see https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation#Using_quaternions_as_rotations for reference
     Q = quaternion_optical * Q * quaternion_optical.inverse();
     float3 trans{tf_ex.translation[0], tf_ex.translation[1], tf_ex.translation[2]};
 
-    append_static_tf_msg(transform_ts_, trans, Q, _base_frame_id, FRAME_ID(sip));
+    if(profile.stream_type() == RS2_STREAM_LABELED_POINT_CLOUD ||
+       profile.stream_type() == RS2_STREAM_OCCUPANCY)
+    {
+        // e.g. tf_ex will include the 
+        // Extrinsics from "Labeled Point Cloud" to "Depth" which is:
+        // - Rotation matrix from Depth (Optical) to Labeled Point Cloud (ROS)
+        // - Translation vector in the Optical CS
+
+        // We are getting these special streams in ROS CS
+        // so no need for rotation of the axes, this is why we use zero_rot_quaternions
+        // We still use the same translation vector (trans) since it is already in the optical CS
+        // and inside append_static_tf_msg we always convert translation to ROS CS
+        append_static_tf_msg(transform_ts_, trans, zero_rot_quaternions, _base_frame_id, FRAME_ID(sip));
+    }
+    else
+    {
+        // All other streams are Optical streams, so we use the "Q" quaternion
+        // Also here, the translation vector is in the Optical CS, and we convert it to ROS CS
+        // inside append_static_tf_msg
+        append_static_tf_msg(transform_ts_, trans, Q, _base_frame_id, FRAME_ID(sip));
+    }
 
-    // Transform stream frame to stream optical frame and publish it
+    // Transform stream frame from ROS CS to optical CS and publish it
+    // We are using zero translation vector here, since no translation between frame and optical_frame, but only rotation
     append_static_tf_msg(transform_ts_, zero_trans, quaternion_optical, FRAME_ID(sip), OPTICAL_FRAME_ID(sip));
```

Comment:
We need 3 calls to ` append_static_tf_msg`?
if ...
   ` append_static_tf_msg`
else
   ' append_static_tf_msg'

' append_static_tf_msg'
?

Reply:
We expose two frames for each sensor
ROS Frame
Optical Frame

The first if/else, includes a append_static_tf_msg that is exposing the ROS_FRAME
the second append_static_tf_msg is exposing the OPTICAL_FRAME (is just a rotation of the ROS_FRAME, with zero translation)